### ISSUE TITLE
Add `yum` section to `extras`

### ISSUE THREAD

**Issue by jakirkham** [conda-forge/conda-forge.github.io#82]
We should come up with a syntax to specify `yum` dependencies of packages in their `meta.yaml` without breaking standard recipe usage. These should be installed as part of the build process in `staged-recipes` and in `*-feedstocks`. The goal is to make these system requirements more visible and properly differentiate which are required for one recipe/package versus another.

**Comment by jakirkham** [comment]
@pelson, I have chosen to reply to [your comment]( https://github.com/conda-forge/pyopengl-feedstock/issues/1#issuecomment-209245395 ) here to avoid noise on the feedstock and also to draw your attention to this issue, which seems like what we are both wanting to discuss anyways. :wink: 

So, I think `meta.yaml` is better. This is why if one is working on a recipe as staged-recipes. They need some way to document the `yum` requirements and use them for testing. The recipe is all they have to work with. When this goes off to a feedstock this information stays intact and is available for use in the first deployment. Further this allows `meta.yaml` to be the singular document of installation requirements. Thus one can see both conda requirements that will be installed and system requirements they will need to address themselves.

However, there is no way for the user to do this in staged-recipes `conda-forge.yml` so they must hack this into the staged-recipes startup. This is undesirable for two main reasons. First, when the feedstock is created it can't deploy on Linux immediately as one needs to go in and edit `conda-forge.yml`. This situation can become even worse if the feedstock generation didn't go smoothly due to some error. Then one must wait for this to get fixed and then make the change. Second, hacking in these `yum` requirements into staged-recipes pollutes the environments for later builds. This makes it hard to tell why a build later fails in the feedstock. It could also create unique errors that occur in staged-recipes due to linkages to system libraries instead of intended packages.

In short, adding this to the `meta.yaml` keeps deployment simple, make specifications clear, documents them in a single place, and avoids contamination of the staged-recipe testing environment.

**Comment by pelson** [comment]
I'm going to re-read the last comment, but initial reaction: I don't think we want to be installing extra yum dependencies in staged-recipes anyway. Unless there are *many* recipes waiting in the wings to do this, I think it is totally reasonable for a recipe to be proposed that has a ```build/skip: True``` defined unconditionally. Only when that has been turned into a feedstock should that be removed, and the appropriate annotation put on the metadata regarding ```yum```.

**Comment by pelson** [comment]
>  they must hack this into the staged-recipes startup

That is something that we both agree is totally unacceptable.

The only sticking point is whether we make staged-recipes deal with the extra yum dependencies or not. My previous comment indicated an inclination not to add extra yum dependencies in staged-recipes - this is only from a desire to reduce the amount of work on us to add, document and maintain such functionality. I think we both agree that adding extra yum dependencies for recipes is something we really want to minimise the use of - the OpenGL question paints us into a corner which yum gets us out of, but how many other corners are there? The effort of adding the extra dependencies in the feedstock is trivial compared to managing the problem on the "many recipes" staged-recipes repo (e.g. what happens when two recipes are queued up, but only one specifies extra yum deps). It is for that reason, I'm OK with the feedstock being brought into existence through staged-recipes without it ever having been built.

**Comment by jakirkham** [comment]
We might not get such an option when it comes to creating a working recipe. Take Qt, for instance, it will need some X11 dependencies from `yum` to just build correctly. Anything that uses the GPU will also have this issue. These are only a few of the cases where this happens. However, the few it does happen for tend to be the more complex packages to build. Having these more challenging packages tested thoroughly and ready for release comes with little burden to the reviewers or maintainers.

However, not ensuring those cases work out of the gate comes at a cost. It puts reviewers in a situation where they now must divide attention between staged-recipes to feedstocks. This ends up being more work to review as one needs to cross-reference all related issues and remain cognizant about the differences between the feedstocks and staged-recipes. Personally, it doesn't seem to me that the CI queues are such a big issue at this point and these [recent]( https://github.com/conda-forge/staged-recipes/pull/332 ) [additions]( https://github.com/conda-forge/staged-recipes/pull/329 ) (almost unnecessary) will only make them less so. The problem at this point is not enough reviewers. Especially as more people realize how awesome conda-forge is. Not having a systematic way of dealing with `yum` further disperses this reviewer resource.

As far as the amount of `yum` installs, I don't think the recipes that are doing these are doing so many. Qt related packages have a few X11 dependencies (that we already install). Things that depend on OpenGL have a Mesa OpenGL and a couple of things that come with it. Honestly, these have been quite fast to install while we have embedded them in staged-recipes. By being able to split them out into the recipe definition, we will avoid these installs most of the time as they are not needed for every package. The few times we do them there won't be so much of a burden.

Concerning the burden of writing this code, won't it be about the same to have it one place versus the other? We still need to parse it and do something with it in either case. If the differences between staged-recipes and feedstocks are growing, we will want to address that as well as it is bad for testing. Ultimately, having this as part of the recipe will keep things quite clean.

Though if you are still not convinced, we could consider adding a new file to the recipe directory just for `yum`. This could be designed to be especially easy to parse. While it would be nice to have this all in `meta.yaml`, it is true this doesn't quite relate to it and maybe introducing a new file for this purpose is reasonable.

Finally, the many recipe PRs. These often tend to be grouped with related packages (a bunch of Django extension packages, a bunch of HDF5 dependent packages, etc.). So, if there is a case where one of them needs `yum` dependencies, they often all need them. In short, I am less worried about one package polluting the other with these requirements. If there is such a rare case, we can always ask the person proposing these additions to split them out as we see fit to find and eliminate such a problem. In practice, I am not sure that this is such a big issue.

**Comment by pelson** [comment]
By the way, you may not have spotted, but conda-smithy already has the ability to use a different image for a feedstock based on the configuration of ```conda-forge.yml```: https://github.com/conda-forge/conda-smithy/blob/master/conda_smithy/configure_feedstock.py#L170

In honesty, that whole ```configure_feedstock.py``` needs a good tidy up (it was hacked together in a few 5 A.M. sessions 😴 💤  ) and the responsibilities better abstracted (without needing to change the feedstocks themselves).

> Concerning the burden of writing this code, won't it be about the same to have it one place versus the other? We still need to parse it and do something with it in either case.

The difference is when we do the parsing. In a staged-recipes context, we need to do that at run time, but in the feedstock context, we have the leisure of doing it at conda-smithy rendering time. It is for that reason I have suggested we put the content in the ```conda-forge.yml``` - by changing the content, we are fundamentally changing *how the feedstock is rendered* by conda-smithy.

As I said earlier, I can certainly see arguments each way on *where* the information lives. Additionally, there is the reproducibility aspect to conda-forge - right now, it is not possible to reproduce the *exact* binary builds that a feedstock produces without knowing exactly what docker image was used, and which version of the dependencies were installed by conda, so I don't think we should use that as an argument for putting the ```yum``` dependencies in the ```meta.yaml```.

You've made some excellent points here which I'd like to discuss in a conda-forge hangout (Friday, though I haven't yet sent the reminder). In order to get us off the fence, I think it is supportable to get a conda-smithy which takes the information from the ```conda-forge.yml``` to get us rolling. Hopefully we can come up with a long-term solution on Friday. Is that OK with you?

**Comment by jakirkham** [comment]
The staged-recipe proposal can be seen in this PR ( https://github.com/conda-forge/staged-recipes/pull/346 ). A sister proposal will be made for `conda-smithy`.

**Comment by jakirkham** [comment]
The sister proposal for `conda-smithy` can be found in this PR ( https://github.com/conda-forge/conda-smithy/pull/135 ).

**Comment by jakirkham** [comment]
Am closing as this has already happened. Please see PR ( https://github.com/conda-forge/conda-smithy/pull/134 ) for the implementation in `conda-smithy`. The implementation for staged-recipes, which carries over to the feedstocks, can be seen with PR ( https://github.com/conda-forge/staged-recipes/pull/371 ). An example use case of this implementation can be seen in PR ( https://github.com/conda-forge/staged-recipes/pull/346 ). Thanks @pelson for working with me to make this happen.
